### PR TITLE
Rolled over courses are closed by default and automatically opened on publish

### PR DESCRIPTION
**Pull request by inulty-dfe**
### Context

When a course is rolled over, it is rolled over to the “rolled over” state.

It is currently rolled over with the same `application_status` it had in the previous cycle.

It doesn’t make sense from a data analysis point of view for the course to be `unpublished` and `open` at the beginning of the cycle.

Therefore we should roll over courses and set the course `application_status` to closed until it is `published`.

We automatically open a course during the publish step to avoid the multi-step process of 'publish', 'open', 'confirm'.

Now the provider needs only to publish a closed course and it will be open for applications pending it's scheduled open date.

### Changes proposed in this pull request

When a provider publishes a course, they must "Open" the course, even if the course was open in the previous cycle.
We've changed this to:

 - Rollover all courses closed
 - automatically open the course when it is published

## Before

|Rolled Over|Published|
|---|---|
|1_unpublished|2_published|

|Opening|Opened|
|---|---|
|3_opening_course|4_opened|

## After

[Screencast from 21-06-24 16:34:43.webm](https://github.com/DFE-Digital/publish-teacher-training/assets/135042929/98276119-3935-413b-bd66-a22c6d761b9f)

### Checklist

- [ ] Make sure all information from the Trello card is in here
- [ ] Attach to Trello card
- [ ] Rebased main
- [ ] Cleaned commit history
- [ ] Tested by running locally
- [ ] Inform data insights team due to database changes

### CODE REVIEW

**Comment by inulty-dfe** [comment]
[`application_status` is rolled over with previous years value](https://trello.com/c/I3QA3Mwb/1428-applicationstatus-is-rolled-over-with-previous-years-value)

**Review by tomas-stefano — COMMENTED** [review]
From the code point of view it looks good. It is unclear if there is any impact on rollover for a course being closed?

File: spec/services/courses/copy_to_provider_service_spec.rb [hunk 1]
```diff
@@ -87,6 +87,20 @@
     end
   end
 
+  context 'when the original course is open' do
+    let(:new_recruitment_cycle) { create(:recruitment_cycle, :next, application_start_date: '2023-09-01') }
```
Comment (avinhurry) [inline spec/services/courses/copy_to_provider_service_spec.rb:91]:
```suggestion
    let(:new_recruitment_cycle) { create(:recruitment_cycle, :next) }
```

**Review by avinhurry — APPROVED** [review]

**Comment by inulty-dfe** [comment]
> From the code point of view it looks good. It is unclear if there is any impact on rollover for a course being closed?

I've illustrated the expected steps for providers to take when launching a course in the new cycle:

1. Publish
2. Open

Every course will require step 2 now because all copied courses will be copied closed.

File: spec/services/courses/copy_to_provider_service_spec.rb [hunk 2]
```diff
@@ -87,6 +87,20 @@
     end
   end
 
+  context 'when the original course is open' do
+    let(:new_recruitment_cycle) { create(:recruitment_cycle, :next) }
```
Comment (avinhurry) [inline spec/services/courses/copy_to_provider_service_spec.rb:91]: I don't think this line is needed

**Review by avinhurry — APPROVED** [review]
Looks good. 

Can we update the PR description 🙏🏼 

> When a provider publishes a course, they must "Open" the course, even if the course was open in the previous cycle.

**Pull request merged**